### PR TITLE
Add sleep API documentation

### PR DESCRIPTION
**Pull request by 0xc0170**
- sleep API
- sleep manager (new in mbed OS 5.6)

This depends on https://github.com/ARMmbed/mbed-os/pull/4912 , as this one gets to 5.6, this should be part of 5.6 documentation. I noticed all PR are to the new_engine branch, this also then.

The aim is to describe how sleep API should work, new API for sleep manager API (use `DeepSleepLock` class in applications). 

An example there was inspired based on SPI changes we did, we can either point directly to the file, or keep this, or any other suggestions?

### CODE REVIEW

**Comment by 0xc0170** [comment]
There's https://github.com/ARMmbed/Handbook/pull/209, similar. This one though is about API reference,  the other one is for contributing (portin sleep hal), I would assume.

**Comment by 0xc0170** [comment]
cc @c1728p9 @bulislaw @sg-

**Comment by bulislaw** [comment]
Looks good, but that should be against the 5.5 branch as we are merging sleep manager for 5.6.

CC: @AnotherButler

**Comment by AnotherButler** [comment]
@0xc0170 I've made some copy edits. Please review and make sure I didn't change the content. Also, this is the right branch.

**Comment by bulislaw** [comment]
@AnotherButler that means we are switching to the new engine for 5.6?

**Comment by 0xc0170** [comment]
I updated the PR, removed sleep API changes taht will come later. 

@AnotherButler What branch should I use? can you please edit and switch it? I'll be away for couple of days, and this should get in. feel free to edit,  to make it integrated.

**Comment by AnotherButler** [comment]
@bulislaw Yes. That is correct.

Also, we're standardizing our APIs. If this is an API and not related to porting, we need to link to the class reference Doxy and have examples in repos instead of hardcoded into the markdown. Can you help with this?

**Comment by bulislaw** [comment]
I'm not sure how to reference a doxy, I'm sure I could figure it, but this wasn't released yet so it probably doesn't have a doxy online.

**Comment by AnotherButler** [comment]
OK. I'll figure out how to build the Doxy, and then I'll embed the code snippet from the Doxy. Do you have time to update the example code, so it's not hardcoded into the markdown?

**Comment by bulislaw** [comment]
But the example doesn't exists as a stand alone program (it's not mbed-os-example-XXX), it's just two snippets of relevant code from mbed os internals. We could potentially link to this file, but I think it wouldn't be as clear, not to mention that the file is evolving all the time so we will have issues maintaining that.

**Comment by bulislaw** [comment]
The example is abstract, it's not an example app in stand alone sense. It's not something you can use out of context and you can't import it anywhere, it's just an illustration of how this sequence of calls should look like. I don't want to include specific driver's code as it's internals can change and things can be moved around, but this hardcoded abstract example will always be correct.

**Comment by AnotherButler** [comment]
@sg- I think this doc should make the standardization we've been working on, or we should move it to a different location because I think our APIs should all have importable examples.

**Comment by bulislaw** [comment]
I would say our examples should be useful in the first place if it makes sense to have them as programs than I agree, in this case it's example of usage from perspective of internal driver.
This PR is based on what @0xc0170 proposed, Martin what do you think? Does it make sense to create dummy example?

**Comment by AnotherButler** [comment]
@bulislaw Don't worry about it in this case. In the future, please try to avoid hardcoding any examples in markdown that may require updating. Is this PR ready to merge, or does it still require content review/changes?

**Comment by bulislaw** [comment]
I think it's ready.

**Comment by AnotherButler** [comment]
@bulislaw We need to embed the class reference Doxy for this before I merge. Do you know the link?

**Comment by bulislaw** [comment]
Search on online doxy doesn't give me back any results, does it need to be regenerated? We are looking for these headers: https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_sleep.h https://github.com/ARMmbed/mbed-os/blob/master/platform/DeepSleepLock.h

**Comment by AnotherButler** [comment]
Perfect, thanks. I can find what should be the right link from there. I'll embed it and merge.

File: docs/refs/api/platform/sleep.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+- `Timeout`
+- `Timer`
+- `SPI`
+- `I2C`
+- `CAN`
+- `SerialBase`
+
+### API
+
+[![View code](https://www.mbed.com/embed/?type=library)](https://docs.mbed.com/docs/mbed-os-api/en/mbed-os-5.6/api/DeepSleepLock_8h_source.html)
```
Comment (bulislaw) [inline docs/refs/api/platform/sleep.md]: I think we should also reference this header:
https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_sleep.h

**Comment by bulislaw** [comment]
Looks great, thanks!

**Pull request merged**